File: csc/2015/bav/bav_1_java/src/ru/compscicenter2015/concurrency/MyFuture.java
package ru.compscicenter2015.concurrency;

import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

public class MyFuture<V> implements Future<V> {
	private final Callable<V> task;
	private volatile Thread currentThread;
	private volatile V result;
	private volatile Throwable exception;
	private final AtomicInteger state;
	
	private static final int NEW       = 0;
	private static final int RUNNING   = 1;
	private static final int CANCELED  = 2;
	private static final int DONE      = 3;
	private static final int ERROR     = 4;
	
	public MyFuture(Callable<V> task) {
		if (task == null)
			throw new NullPointerException();
		this.task = task;
		exception = null;
		result = null;
		state = new AtomicInteger(NEW);
	}

	public void start() {
	//	System.out.println("Im ready to do smth usefull");
		state.compareAndSet(NEW, RUNNING);
		currentThread = Thread.currentThread();
		try {
			result = task.call();
		} catch (Throwable e) {
			state.compareAndSet(RUNNING, ERROR);
			exception = e; 
		} 
		state.compareAndSet(RUNNING, DONE);
		synchronized (task) {
			task.notify();	
		}
	}
	
	@SuppressWarnings("finally")
	@Override
	public boolean cancel(boolean mayInterruptIfRunning) {
		if (state.get() == DONE)
			return false;
		if (state.get() == CANCELED)
			return true;		
		if (state.compareAndSet(NEW, CANCELED)) 
			return true;
		if (mayInterruptIfRunning) {
			try {
				Thread t = null;
				while ((t = currentThread) == null)
					Thread.yield();
				if (t != null) {
					t.interrupt();
				}
			} finally {
				state.set(CANCELED);
			}
		}
		return false;
	}

	@Override
	public boolean isCancelled() {
		return state.get() == CANCELED;
	}

	@Override
	public boolean isDone() {
		return state.get() == DONE;
	}

	@Override
	public V get() throws InterruptedException, ExecutionException {
		synchronized (task) {
			while (state.get() == NEW || state.get() == RUNNING) {
				task.wait();
			}
		}
		if (state.get() == ERROR)
			throw new ExecutionException(exception);
		if (state.get() == CANCELED)
			throw new CancellationException();
		
		return result;
	}

	@Override
	public V get(long timeout, TimeUnit unit) throws InterruptedException,
			ExecutionException, TimeoutException {
		if (unit == null)
			throw new NullPointerException();
		long deadline = unit.toMillis(timeout) + System.currentTimeMillis();
		synchronized (task) {
			while (System.currentTimeMillis() < deadline && (state.get() == NEW || state.get() == RUNNING)) {
				task.wait(unit.toMillis(timeout));
			}
		}
		if (state.get() == ERROR)
			throw new ExecutionException(exception);
		if (state.get() == CANCELED)
			throw new CancellationException();
		if (state.get() != DONE)
			throw new TimeoutException();
		return result;
	}
	
	public String getState() {
		switch(state.get()) {
		case NEW:
			return "New";
		case RUNNING:
			return "Running";
		case CANCELED:
			return "Canceled";	
		case DONE:
			return "Done";
		case ERROR:
			return "Error";
		}
		return "";
	}
}

File: csc/2015/any/any_1_kotlin/src/ru/bronti/hpcource/hw1/MyFuture.kt
package ru.bronti.hpcource.hw1

import java.util.LinkedList
import java.util.Queue
import java.util.concurrent.Callable
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.RejectedExecutionException
import java.lang.ThreadGroup
import java.util.concurrent.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

public class MyFuture<V> (private val task: Callable<V>,
                          val name: String,
                          private val threadPool: MyThreadPool): Future<V> {

    private val resultMonitor = Object()

    val WAITING = 0
    val RUNNING = 1
    val DONE    = 2
    val CANCELLED = 4

    val status = AtomicInteger(WAITING)

    volatile var workerThread: MyThread? = null
    volatile var result: V = null
    volatile var exeption: Exception? = null

    override public fun isDone(): Boolean = status.get() % CANCELLED >= DONE
    override public fun isCancelled(): Boolean = status.get() >= CANCELLED

    private fun doWaitOnce(expirationTime: Long?): Boolean {
        if (Thread.currentThread() is MyThread &&
                !threadPool.isShut &&
                !threadPool.queue.isEmpty() &&
                !isCancelled()) {
            //println(Thread.currentThread().getName() + " is waiting for " + name)
            val wasInterrupted = Thread.interrupted()
            (Thread.currentThread() as MyThread).task.runTask(false)
            if (wasInterrupted) {
                Thread.currentThread().interrupt()
            }
        }
        else {
            synchronized(resultMonitor) {
                if (!isCancelled()) {
                    if (expirationTime != null) {
                        resultMonitor.wait(expirationTime - System.currentTimeMillis())
                    } else {
                        resultMonitor.wait()
                    }
                }
            }
        }
        if (expirationTime != null && System.currentTimeMillis() >= expirationTime) {
            return true
        }
        return false
    }

    private fun doGet(expirationTime: Long?): V {
        var timedOut = false
        while (!isDone() && !isCancelled() && !timedOut) {
            timedOut = doWaitOnce(expirationTime)
        }
        if (isCancelled()) {
            throw CancellationException()
        }
        if (!timedOut) {
            if (exeption != null) {
                throw ExecutionException(exeption)
            }
            return result
        }
        else {
            throw TimeoutException()
        }
    }

    override public fun get(timeout: Long, unit: TimeUnit): V {
        val expirationTime = System.currentTimeMillis() + unit.toMillis(timeout);
        return doGet(expirationTime)
    }

    override public fun get(): V {
        return doGet(null)
    }

    override public fun cancel(mayInterruptIfRunning: Boolean): Boolean {
        var result: Boolean = false
        synchronized(resultMonitor) {
            status.set(status.get() % CANCELLED + CANCELLED)
            resultMonitor.notifyAll()
        }
        if (mayInterruptIfRunning) {
            if (status.get() % CANCELLED == RUNNING) {
                workerThread!!.interrupt()
                result = true
            }
            //println(name + " interrupted while running")
        }
        result = result or (status.get() % CANCELLED == WAITING)
        status.set(CANCELLED + DONE + status.get() % DONE)
        return result
    }

    throws(javaClass<InterruptedException>())
    fun run(thread: MyThread) {
        workerThread = thread
        var temp: V = null
        if (status.compareAndSet(WAITING, RUNNING)) {
            //println(name + " running")
            try {
                temp = task.call()
            } catch (e: InterruptedException) {
                synchronized(resultMonitor) {
                    if (status.compareAndSet(RUNNING, RUNNING + DONE)) {
                        result = temp
                        resultMonitor.notifyAll()
                        throw e
                    }
                }
            } catch (e: Exception) {
                exeption = e
            }
            if (!isCancelled()) {
                synchronized(resultMonitor) {
                    if (status.compareAndSet(RUNNING, RUNNING + DONE)) {
                        result = temp
                        resultMonitor.notifyAll()
                    }
                }
            }
        }
        else {
            //println("skipped execution by " + name)
        }
    }

    fun getTaskAndCancel(): Callable<V> {
        cancel(true)
        return task
    }
}